components/org.wso2.carbon.metrics.data.service/src/main/java/org/wso2/carbon/metrics/data/service/dao/ReporterDAO.java

```diff
@@ -176,6 +176,64 @@ private void validateMetricAttributes(MetricType metricType, List<MetricAttribut
         return results;
     }
 
+    public Map<String, MetricType> queryHierarchicalMetrics(String source, String path) {
+        Map<String, MetricType> results = new HashMap<>();
+        Connection connection = null;
+        try {
+            connection = dataSource.getConnection();
+            for (MetricType metricType : MetricType.values()) {
+                List<String> names = getHierarchicalMetricNames(connection, source, path, metricType);
+                if (!names.isEmpty()) {
+                    for (String name : names) {
+                        results.put(name, metricType);
+                    }
+                }
+            }
+            connection.close();
```

Comment:
Closing a connection needs to happen in the finally block. In cases if there are any exception thrown the obtained connection may not be closed, and it'll cause the connection leak.